## Importación de Librerías iniciales:  

In [24]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

### Establecer una semilla aleatoria

In [25]:
seed = 42

#### Lectura de base de datos:

In [26]:
data = pd.read_csv(r'C:\Users\dg_su\OneDrive - Universidad Tecnologica del Peru\Estudios\Cursos\Analisis de Datos\Programa Data Science for Business\codigo\base de datos\TelcoCustomerChurn.csv')

In [27]:
data.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [28]:
data.shape

(7043, 21)

### Definimos el target y las covariables:

In [29]:
id_columns = 'customerID'
target = 'Churn'

###  Convertir valores de tipo texto a binario:

In [30]:
data[target].value_counts()

Churn
No     5174
Yes    1869
Name: count, dtype: int64

In [31]:
data[target] = data[target].map({'Yes':1,'No':0})

In [32]:
data[target].value_counts()

Churn
0    5174
1    1869
Name: count, dtype: int64

In [33]:
data[target].value_counts(normalize=True)

Churn
0    0.73463
1    0.26537
Name: proportion, dtype: float64

Se puede apreciar que el 73% de los datos de la variable objetivo son 0, mientras que 26.5% son 1

In [34]:
data[target].mean()

0.2653698707936959

### Se procede a eliminar la columna customerID, ya que no proporciona información relevante

In [35]:
data = data.drop('customerID',axis=1)

In [36]:
data.head()

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  Female              0     Yes         No       1           No   
1    Male              0      No         No      34          Yes   
2    Male              0      No         No       2          Yes   
3    Male              0      No         No      45           No   
4  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity OnlineBackup  \
0  No phone service             DSL             No          Yes   
1                No             DSL            Yes           No   
2                No             DSL            Yes          Yes   
3  No phone service             DSL            Yes           No   
4                No     Fiber optic             No           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No          No          No              No  Month-to-month   
1              Yes          No          No              No        One year   
2               No          No          No              No  Month-to-month   
3              Yes         Yes          No              No        One year   
4               No          No          No              No  Month-to-month   

  PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  \
0              Yes           Electronic check           29.85        29.85   
1               No               Mailed check           56.95       1889.5   
2              Yes               Mailed check           53.85       108.15   
3               No  Bank transfer (automatic)           42.30      1840.75   
4              Yes           Electronic check           70.70       151.65   

   Churn  
0      0  
1      0  
2      1  
3      0  
4      1

## Division en conjutos de entrenamiento y prueba

In [37]:
X = data.drop(target, axis = 1)
y = data[target]

In [38]:
X.head()

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  Female              0     Yes         No       1           No   
1    Male              0      No         No      34          Yes   
2    Male              0      No         No       2          Yes   
3    Male              0      No         No      45           No   
4  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity OnlineBackup  \
0  No phone service             DSL             No          Yes   
1                No             DSL            Yes           No   
2                No             DSL            Yes          Yes   
3  No phone service             DSL            Yes           No   
4                No     Fiber optic             No           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No          No          No              No  Month-to-month   
1              Yes          No          No              No        One year   
2               No          No          No              No  Month-to-month   
3              Yes         Yes          No              No        One year   
4               No          No          No              No  Month-to-month   

  PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  
0              Yes           Electronic check           29.85        29.85  
1               No               Mailed check           56.95       1889.5  
2              Yes               Mailed check           53.85       108.15  
3               No  Bank transfer (automatic)           42.30      1840.75  
4              Yes           Electronic check           70.70       151.65

In [39]:
y.head()

0    0
1    0
2    1
3    0
4    1
Name: Churn, dtype: int64

In [40]:
X.shape

(7043, 19)

In [41]:
y.shape

(7043,)

In [42]:
# función para dividir la base en piloto y control
from sklearn.model_selection import train_test_split

In [43]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2, stratify= y, random_state = seed)

In [44]:
X_train.head()

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
3738    Male              0      No         No      35           No   
3151    Male              0     Yes        Yes      15          Yes   
4860    Male              0     Yes        Yes      13           No   
3867  Female              0     Yes         No      26          Yes   
3810    Male              0     Yes        Yes       1          Yes   

         MultipleLines InternetService OnlineSecurity OnlineBackup  \
3738  No phone service             DSL             No           No   
3151                No     Fiber optic            Yes           No   
4860  No phone service             DSL            Yes          Yes   
3867                No             DSL             No          Yes   
3810                No             DSL             No           No   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
3738              Yes          No         Yes             Yes  Month-to-month   
3151               No          No          No              No  Month-to-month   
4860               No         Yes          No              No        Two year   
3867              Yes          No         Yes             Yes        Two year   
3810               No          No          No              No  Month-to-month   

     PaperlessBilling            PaymentMethod  MonthlyCharges TotalCharges  
3738               No         Electronic check           49.20      1701.65  
3151               No             Mailed check           75.10      1151.55  
4860               No             Mailed check           40.55       590.35  
3867              Yes  Credit card (automatic)           73.50       1905.7  
3810               No         Electronic check           44.55        44.55

In [45]:
y_train.head()

3738    0
3151    0
4860    0
3867    0
3810    0
Name: Churn, dtype: int64

In [46]:
X_train.shape, X_test.shape

((5634, 19), (1409, 19))

In [47]:
y_train.shape, y_test.shape

((5634,), (1409,))

In [48]:
y.mean(), y_train.mean(), y_test.mean()

(0.2653698707936959, 0.2653532126375577, 0.2654364797728886)

### ANÁLISIS EXPLORATORIO

In [49]:
X_train.columns

Index(['gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure',
       'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity',
       'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV',
       'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod',
       'MonthlyCharges', 'TotalCharges'],
      dtype='object')

In [50]:
X_train.dtypes

gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges         object
dtype: object

In [51]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5634 entries, 3738 to 5639
Data columns (total 19 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   gender            5634 non-null   object 
 1   SeniorCitizen     5634 non-null   int64  
 2   Partner           5634 non-null   object 
 3   Dependents        5634 non-null   object 
 4   tenure            5634 non-null   int64  
 5   PhoneService      5634 non-null   object 
 6   MultipleLines     5634 non-null   object 
 7   InternetService   5634 non-null   object 
 8   OnlineSecurity    5634 non-null   object 
 9   OnlineBackup      5634 non-null   object 
 10  DeviceProtection  5634 non-null   object 
 11  TechSupport       5634 non-null   object 
 12  StreamingTV       5634 non-null   object 
 13  StreamingMovies   5634 non-null   object 
 14  Contract          5634 non-null   object 
 15  PaperlessBilling  5634 non-null   object 
 16  PaymentMethod     5634 non-null   object 
 1

In [52]:
X_train.describe()

SeniorCitizen       tenure  MonthlyCharges
count    5634.000000  5634.000000     5634.000000
mean        0.163294    32.485091       64.929961
std         0.369667    24.568744       30.138105
min         0.000000     0.000000       18.400000
25%         0.000000     9.000000       35.662500
50%         0.000000    29.000000       70.500000
75%         0.000000    55.000000       90.000000
max         1.000000    72.000000      118.750000